### PR TITLE
AWSManagedMachinePool - Launch Template support

### LINKED ISSUE #2055: AWSManagedMachinePool - Launch Template support
/kind feature

**Describe the solution you'd like**
The managed EKS nodegroups backing AWSManagedMachinePool [support launch templates](https://aws.amazon.com/blogs/containers/introducing-launch-template-and-custom-ami-support-in-amazon-eks-managed-node-groups/). This functionality was put on hold until both the `AWSMachinePools` and `AWSManagedMachinePools` support was merged.

**Anything else you would like to add:**
Discussed in https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1916#issuecomment-700741381

### PR DESCRIPTION
**Pull request by richardchen331**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add launch template support to `AWSManagedMachinePool`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2055

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Add launch template support to `AWSManagedMachinePool`.
```

### CODE REVIEW

**Comment by sedefsavas** [comment]
@richardchen331 @richardcase is this implementation in line with https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3024?

**Comment by richardcase** [comment]
> @richardchen331 @richardcase is this implementation in line with #3024?

@sedefsavas - i need to update the proposal based on a chat that i had with @richardchen331. We decided on a way forward to unblock development. If after the proposal has been updated and changes are needed we can go back and refactor. I will check the proposal updated by EOD Wednesday.

File: exp/api/v1beta1/awsmanagedmachinepool_types.go [hunk 1]
```diff
@@ -144,6 +143,9 @@ type AWSManagedMachinePoolSpec struct {
 	// +kubebuilder:default:=onDemand
 	// +optional
 	CapacityType *ManagedMachinePoolCapacityType `json:"capacityType,omitempty"`
+
+	// +optional
+	AWSLaunchTemplate *AWSLaunchTemplate `json:"awsLaunchTemplate,omitempty"`
```
Comment (richardcase) [inline exp/api/v1beta1/awsmanagedmachinepool_types.go:148]: It would be good to have a comment here. Maybe worth mentioning that you can't use this if you fields X, Y, Z?

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 2]
```diff
@@ -182,6 +201,21 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
...
+			machinePoolScope.Error(err, "failed to reconcile launch template")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.reconcileTags(machinePoolScope, ec2Scope); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "error updating tags")
+		}
+
+		// set the LaunchTemplateReady condition
+		conditions.MarkTrue(machinePoolScope.ManagedMachinePool, expinfrav1.LaunchTemplateReadyCondition)
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:216]: We should probably mark this condition false in the above error scenarios, along with a reason.

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 3]
```diff
@@ -192,15 +226,40 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
...
+
+		if launchTemplate == nil {
+			machinePoolScope.V(2).Info("Unable to locate launch template")
+			r.Recorder.Eventf(machinePoolScope.ManagedMachinePool, corev1.EventTypeNormal, "NoLaunchTemplateFound", "Unable to find matching launch template")
+			controllerutil.RemoveFinalizer(machinePoolScope.ManagedMachinePool, expinfrav1.ManagedMachinePoolFinalizer)
+			return ctrl.Result{}, nil
+		}
+
+		machinePoolScope.Info("deleting launch template", "name", launchTemplate.Name)
+		if err := ec2Svc.DeleteLaunchTemplate(*launchTemplateID); err != nil {
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:255]: Should the delete use the `launchTemplate.LaunchTemplateId` instead?

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 4]
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
 		return results
 	}
 }
+
+func (r *AWSManagedMachinePoolReconciler) getEC2Service(scope scope.EC2Scope) services.EC2MachineInterface {
+	return ec2.NewService(scope)
+}
+
+func (r *AWSManagedMachinePoolReconciler) reconcileLaunchTemplate(machinePoolScope *scope.ManagedMachinePoolScope, ec2Scope scope.EC2Scope) error {
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:333]: Is there anyway we could have a common function for this across [AWSMachinePool](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/833ca27c87eab8a8f7fd98ca85b46dcd1a22e94f/exp/controllers/awsmachinepool_controller.go#L383) and AWSManagedMachinePool?

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 5]
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
...
+			return err
+		}
+		machinePoolScope.SetLaunchTemplateVersionStatus(version)
+		return machinePoolScope.PatchObject()
+	}
+
+	return nil
+}
+
+func (r *AWSManagedMachinePoolReconciler) reconcileTags(machinePoolScope *scope.ManagedMachinePoolScope, ec2Scope scope.EC2Scope) error {
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:425]: Same comment about reusing the logic between the 2 different machine pool controllers

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 6]
```diff
@@ -266,3 +325,187 @@ func managedControlPlaneToManagedMachinePoolMapFunc(c client.Client, gvk schema.
...
+	if err != nil {
+		return err
+	}
+	if tagsChanged {
+		r.Recorder.Eventf(machinePoolScope.ManagedMachinePool, corev1.EventTypeNormal, "UpdatedTags", "updated tags on resources")
+	}
+	return nil
+}
+
+func (r *AWSManagedMachinePoolReconciler) machinePoolAnnotationJSON(machinePool *expinfrav1.AWSManagedMachinePool, annotation string) (map[string]interface{}, error) {
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:441]: This and the `machinePoolAnnotation` below feels like it could be useful on the scope. wdyt?

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 4 above)
Comment (richardcase) [reply exp/controllers/awsmanagedmachinepool_controller.go:333]: Do we need to extend the ec2 service? Or perhaps refactor and create a launch template service?

**Review by richardcase — CHANGES_REQUESTED** [review]

**Comment by sedefsavas** [comment]
Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.

**Comment by richardcase** [comment]
> Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.

Agree on adding this to the e2e tests for eks. This is covered in the updated proposal. @richardchen331 i can guide you through this if you want?

File: exp/api/v1beta1/awsmanagedmachinepool_types.go [hunk 7]
```diff
@@ -105,7 +105,6 @@ type AWSManagedMachinePoolSpec struct {
 
 	// AMIType defines the AMI type
 	// +kubebuilder:validation:Enum:=AL2_x86_64;AL2_x86_64_GPU;AL2_ARM_64
-	// +kubebuilder:default:=AL2_x86_64
```
[the remark this answers is not in the document]
Comment (richardcase) [reply exp/api/v1beta1/awsmanagedmachinepool_types.go:108]: We still need this default.

File: exp/api/v1beta1/awsmanagedmachinepool_types.go (hunk 1 above)
Comment (richardchen331) [reply exp/api/v1beta1/awsmanagedmachinepool_types.go:152]: Sounds good. Added comment here.

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 2 above)
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go:243]: That makes sense. Added logic to mark the condition to false in this case.

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 3 above)
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go:283]: There's no LaunchTemplateId field in `AWSLaunchTemplate` at this moment, that's why i'm using launchTemplateID instead. Please let me know if that makes sense. Thanks!

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 4 above)
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go]: Sounds good. I moved the logic into the launchtemplate service (`ReconcileLaunchTemplate`).

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 5 above)
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go]: Moved the logic into the launchtemplate service as well.

File: exp/controllers/awsmanagedmachinepool_controller.go (hunk 6 above)
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go]: Sounds good. Moved the logic into launchtemplate scope.

File: exp/api/v1beta1/awsmanagedmachinepool_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (richardchen331) [reply exp/api/v1beta1/awsmanagedmachinepool_types.go]: Got it. Added back the default.

**Comment by richardchen331** [comment]
> > Given that this is an important feature and changes the way we interact with `AWSManagedMachinePool`, we would want to have this in e2e tests and there is already a proposal out but would be good to update existing docs.
> 
> Agree on adding this to the e2e tests for eks. This is covered in the updated proposal. @richardchen331 i can guide you through this if you want?

@richardcase  Agree that we should have e2e tests for it. It would be great if you could guide me on how to add e2e tests for eks :)

**Comment by richardcase** [comment]
> Agree that we should have e2e tests for it. It would be great if you could guide me on how to add e2e tests for eks :)

I'll ping you in slack and we can arrange a zoom sessions

**Comment by pydctw** [comment]
As you created a new test flavor, it should be added to files section in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L179-L207.

**Comment by richardchen331** [comment]
> I was just reviewing this PR 😀 As you have created a new test flavor, it should be added to files section in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L179-L207.

@pydctw Thank you! Just updated the conf and i'll retry the test again :)

**Comment by richardchen331** [comment]
Hi @richardcase @sedefsavas , I added e2e test for this change. Could you help review it again? Thanks!

File: test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  name: "${CLUSTER_NAME}-pool-0"
+spec:
+  clusterName: "${CLUSTER_NAME}"
+  replicas: ${WORKER_MACHINE_COUNT}
+  template:
+    spec:
+      version: "EKS_KUBERNETES_VERSION"
+      clusterName: "${CLUSTER_NAME}"
+      bootstrap:
+        dataSecretName: "${CLUSTER_NAME}-pool-0-userdata"
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml:13]: This would not work here. The existing behavior requires the user to specify a data secret used as bootstrap data in AWSMachinePool (https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/scope/machinepool.go#L126)

File: test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachinePool
+metadata:
+  name: "${CLUSTER_NAME}-pool-0"
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply test/e2e/data/eks/cluster-template-eks-managed-machinepool-with-launch-template-only.yaml]: Got it. I updated the machine pool to a different name, does that sound good?

**Comment by sedefsavas** [comment]
Is this ready for another round of reviews?

**Comment by richardchen331** [comment]
> Is this ready for another round of reviews?

@sedefsavas @richardcase  Yes it is. Could you help take another look? Thanks!

**Comment by richardcase** [comment]
A couple of things to note:

- Added a follow-up issue to add support for the EKS bootstrapper for `AWSManagedMachinePools`. See #3363
- Added a note to handle the deprecated fields in v1beta2. See #2355

File: exp/api/v1beta1/awsmanagedmachinepool_webhook.go [hunk 10]
```diff
@@ -188,6 +214,16 @@ func (r *AWSManagedMachinePool) validateImmutable(old *AWSManagedMachinePool) fi
 	appendErrorIfMutated(old.Spec.AMIType, r.Spec.AMIType, "amiType")
 	appendErrorIfMutated(old.Spec.RemoteAccess, r.Spec.RemoteAccess, "remoteAccess")
 	appendErrorIfSetAndMutated(old.Spec.CapacityType, r.Spec.CapacityType, "capacityType")
+	if (old.Spec.AWSLaunchTemplate != nil && r.Spec.AWSLaunchTemplate == nil) ||
```
Comment (richardcase) [inline exp/api/v1beta1/awsmanagedmachinepool_webhook.go:217]: Could we use the `appendErrorIfMutated` for this test?

File: exp/controllers/awsmachinepool_controller.go [hunk 11]
```diff
@@ -140,6 +139,21 @@ func (r *AWSMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, err
 	}
 
+	// Create the launch template scope
+	launchTemplateScope, err := scope.NewLaunchTemplateScope(scope.LaunchTemplateScopeParams{
```
Comment (richardcase) [inline exp/controllers/awsmachinepool_controller.go:143]: Do we need a new scope at this level? We have the `MachinePoolScope` is the scope of the reconciliation of the machine pool, so personally i don't think we need it here. But there is nothing stopping us creating the scope further down.

Comment (richardcase) [reply exp/controllers/awsmachinepool_controller.go:143]: I would potentially change `ec2Svc.ReconcileLaunchTemplate` so that it accepts the machine pool scope like we do for the [NodegroupService](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/eks/service.go#L72).

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 12]
```diff
@@ -150,15 +164,32 @@ func (r *AWSManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctr
...
 		AllowAdditionalRoles: r.AllowAdditionalRoles,
 		Endpoints:            r.Endpoints,
+		InfraCluster:         managedControlPlaneScope,
 	})
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "failed to create scope")
 	}
 
+	// Create the launch template scope
+	launchTemplateScope, err := scope.NewLaunchTemplateScope(scope.LaunchTemplateScopeParams{
```
Comment (richardcase) [inline exp/controllers/awsmanagedmachinepool_controller.go:174]: Same comment as previous around the introduction of a new scope as this level.

File: pkg/cloud/scope/launchtemplate.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1beta1"
+	expclusterv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+const (
+	// AWSManagedControlPlaneKind is the Kind of AWSManagedControlPlane.
+	AWSManagedControlPlaneKind = "AWSManagedControlPlane"
```
Comment (richardcase) [inline pkg/cloud/scope/launchtemplate.go:39]: We have this defined in the [machine controller](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0babdf15d5584744c4693e8be4ab44c580847531/controllers/awsmachine_controller.go#L79)...do you think we should move it somewhere shared?

File: pkg/cloud/scope/launchtemplate.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+	// Start with the cluster-wide tags...
+	tags.Merge(m.InfraCluster.AdditionalTags())
+	// ... and merge in the Machine's
+	tags.Merge(m.additionalTags)
+
+	return tags
+}
+
+func (m *LaunchTemplateScope) GetRawBootstrapData() ([]byte, error) {
```
Comment (richardcase) [inline pkg/cloud/scope/launchtemplate.go:139]: We have this already on the [MachinePoolScope](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/0babdf15d5584744c4693e8be4ab44c580847531/pkg/cloud/scope/machinepool.go#L126) so it feels like we should using that implementation and passing the result down into the service.

Comment (richardcase) [reply pkg/cloud/scope/launchtemplate.go:139]: Same with the tags above

Comment (richardcase) [reply pkg/cloud/scope/launchtemplate.go:139]: As a general pattern if we have the parent machine pool scope that kicks of reconciliation of another resource we can pass down the scope or the values instead of reimplemnting the functions.

**Comment by richardcase** [comment]
@richardchen331 - just checking on this. How are you egtting on, did you need anything?

**Comment by richardchen331** [comment]
Hi @richardcase , thanks for checking in on this. I'm busy w/ work recently and I'll update the PR based on the feedback this week.

File: exp/api/v1beta1/awsmanagedmachinepool_webhook.go (hunk 10 above)
Comment (richardchen331) [reply exp/api/v1beta1/awsmanagedmachinepool_webhook.go:217]: I was thinking that we should allow user to modify the launch template (e.g. update AMI ID), but should not allow user to change a node group from using launch template to not using launch template (and vice versa). That's why I didn't use `appendErrorIfMutated`. Let me know if that makes sense. Thanks!

File: exp/controllers/awsmachinepool_controller.go (hunk 11 above)
Comment (richardchen331) [reply exp/controllers/awsmachinepool_controller.go:143]: Good suggestion! I refactored the code so we don't create a new scope here and in the AWSMMP controller.

File: pkg/cloud/scope/launchtemplate.go (hunk 13 above)
Comment (richardchen331) [reply pkg/cloud/scope/launchtemplate.go:39]: I don't know what is a good place to put the constant.. Do you have any suggestions?

File: pkg/cloud/scope/launchtemplate.go (hunk 14 above)
Comment (richardchen331) [reply pkg/cloud/scope/launchtemplate.go:139]: That makes sense. Updated the code to reuse implementation from MachinePoolScope

**Comment by richardchen331** [comment]
@richardcase I refactored the code based on your feedback. Could you help take another look? Thanks!

**Comment by richardcase** [comment]
@richardchen331 - sorry for the delay in getting back to this. I will be doing another pass this week.

File: pkg/cloud/scope/launchtemplate.go [hunk 15]
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (richardcase) [inline pkg/cloud/scope/launchtemplate.go:2]:
```suggestion
Copyright 2022 The Kubernetes Authors.
```

File: pkg/cloud/scope/launchtemplate.go (hunk 13 above)
Comment (richardcase) [reply pkg/cloud/scope/launchtemplate.go:32]: Thats a hard one. Perhaps we add it to the current EKS control plane API version, so `controlplane/eks/api/v1beta1`

File: templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+  region: us-west-2
+  eksClusterName: dkc-eks-lt
+  sshKeyName: default
+  version: v1.19
+  roleName: eks-controlplane.cluster-api-provider-aws.sigs.k8s.io
+  iamAuthenticatorConfig:
+    mapRoles:
+      - groups:
+          - system:masters
+        rolearn: arn:aws:iam::707343435239:role/aws-dev_databricks-view-only
```
Comment (richardcase) [inline templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml:35]: Guessing these need to be removed from the template :smiley:

Comment (richardcase) [reply templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml:35]: If you could add the envsubst declarations like the other templates that would be great.

File: test/e2e/suites/managed/machine_pool.go [hunk 17]
```diff
@@ -63,11 +70,25 @@ func ManagedMachinePoolSpec(ctx context.Context, inputGetter func() ManagedMachi
...
-	shared.Byf("creating an applying the %s template", EKSManagedPoolOnlyFlavor)
+	shared.Byf("creating an applying the %s template", input.Flavor)
 	configCluster := input.ConfigClusterFn(input.ClusterName, input.Namespace.Name)
-	configCluster.Flavor = EKSManagedPoolOnlyFlavor
+	configCluster.Flavor = input.Flavor
 	configCluster.WorkerMachineCount = pointer.Int64Ptr(1)
-	err := shared.ApplyTemplate(ctx, configCluster, input.BootstrapClusterProxy)
+	workloadClusterTemplate := shared.GetTemplate(ctx, configCluster)
+	if input.UsesLaunchTemplate {
+		userDataTemplate := `#!/bin/bash
```
Comment (richardcase) [inline test/e2e/suites/managed/machine_pool.go:79]: Why do we need this? Does this not come from the bootstrap secret?

(i will go back and check the code ^^^^)

**Review by richardcase — COMMENTED** [review]
@richardchen331 - a few comments. Would you also be able to rebase and squash your commits so they tell a story?

File: test/e2e/suites/managed/machine_pool.go (hunk 17 above)
Comment (richardcase) [reply test/e2e/suites/managed/machine_pool.go:80]: Created #3363  for this

File: pkg/cloud/scope/launchtemplate.go (hunk 15 above)
Comment (richardchen331) [reply pkg/cloud/scope/launchtemplate.go]: Updated

File: templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml (hunk 16 above)
Comment (richardchen331) [reply templates/cluster-template-eks-managedmachinepool-launchtemplate.yaml]: oops, thanks for catching this! this is not intended and removed now

**Comment by richardchen331** [comment]
@richardcase This PR is rebased and ready for review. Could you help take a look when you have time? Thanks!

**Comment by richardcase** [comment]
Awesome thanks @richardchen331 :tada:  I will review first thing tomorrow morning. And great to see the e2e passing.

**Comment by richardchen331** [comment]
@richardcase could you kindly help take a look at this PR when you have time? it would be difficult to keep rebasing as the master branch changes... Thanks!

**Comment by pydctw** [comment]
@richardchen331, I reviewed the PR a long time ago but can review again. 

Is the PR almost final? Asking because there is a plan to bump API to v1beta2 in near future and I think we should get this one merged prior if possible as this PR make many changes to v1beta1 API.

cc @sedefsavas @richardcase (out this week)

**Comment by richardchen331** [comment]
> @richardchen331, I reviewed the PR a long time ago but can review again.
> 
> Is the PR almost final? Asking because there is a plan to bump API to v1beta2 in near future and I think we should get this one merged prior if possible as this PR make many changes to v1beta1 API.
> 
> cc @sedefsavas @richardcase (out this week)

Thank you @pydctw ! Yes I have addressed all prior review comments already and it's ready for final review. Please feel free to reach out to me if you have any questions about this PR. Thanks again for reviewing!

File: test/e2e/suites/managed/eks_test.go [hunk 18]
```diff
@@ -36,12 +36,13 @@ import (
...
-		clusterName      string
-		cniAddonName     = "vpc-cni"
-		corednsAddonName = "coredns"
+		namespace            *corev1.Namespace
+		ctx                  context.Context
+		specName             = "cluster"
+		clusterName          string
+		cniAddonName         = "vpc-cni"
+		corednsAddonName     = "coredns"
+		eksKubernetesVersion = "v1.21.14"
```
Comment (pydctw) [inline test/e2e/suites/managed/eks_test.go:45]: Can we use [KUBERNETES_VERSION](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/data/e2e_eks_conf.yaml#L177) from e2e_eks_conf.yaml instead of specifying one here?

MachinePool uses k8s version of v1.21.14 while AWSManagedControlPlane uses v1.22. See [artifacts](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/3094/pull-cluster-api-provider-aws-e2e-eks/1557132529587195904/artifacts/clusters/bootstrap/resources/cluster-fjcwdc/). Unless there is a specific reason for it, I think it will be great to be consistent.

File: pkg/cloud/services/ec2/launchtemplate.go [hunk 19]
```diff
@@ -27,15 +28,296 @@ import (
...
+		version, err := ec2svc.GetLaunchTemplateLatestVersion(scope.GetLaunchTemplateIDStatus())
+		if err != nil {
+			return err
+		}
+
+		scope.SetLaunchTemplateLatestVersionStatus(version)
+		return scope.PatchObject()
+	}
+
+	if needsUpdate || tagsChanged || *imageID != *launchTemplate.AMI.ID {
```
Comment (pydctw) [inline pkg/cloud/services/ec2/launchtemplate.go:159]: I am confused by the logic and trying to confirm my understanding. When does this `if` statement run?

`if needsUpdate || tagsChanged || *imageID != *launchTemplate.AMI.ID`  is true, doesn't it return on line 156 and never reaches this line?

**Comment by pydctw** [comment]
Is the plan to mark `AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile` fields as deprecated in this PR and remove in v1beta2? 

That is my understanding from the [proposal](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3365/files#diff-d6d8565c707995a7204010e1acade42238594c2bcf88d13a6ed7bd316369942eR244) but want to confirm.

**Comment by richardcase** [comment]
Yes they should be marked as deprecated. Then when we bump the API version in the future we can remove them.....assuming we have suitable conversion functions in place.

File: exp/controllers/awsmanagedmachinepool_controller.go [hunk 20]
```diff
@@ -198,15 +243,40 @@ func (r *AWSManagedMachinePoolReconciler) reconcileNormal(
...
 
+	if machinePoolScope.ManagedMachinePool.Spec.AWSLaunchTemplate != nil {
+		launchTemplateID := machinePoolScope.ManagedMachinePool.Status.LaunchTemplateID
+		launchTemplate, _, err := ec2Svc.GetLaunchTemplate(machinePoolScope.LaunchTemplateName())
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if launchTemplate == nil {
+			machinePoolScope.V(2).Info("Unable to locate launch template")
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply exp/controllers/awsmanagedmachinepool_controller.go:265]: done

File: pkg/cloud/services/ec2/launchtemplate.go (hunk 19 above)
Comment (richardchen331) [reply pkg/cloud/services/ec2/launchtemplate.go:159]: Thanks for pointing it out! You're right and let me update the code here.

File: test/e2e/suites/managed/eks_test.go (hunk 18 above)
Comment (richardchen331) [reply test/e2e/suites/managed/eks_test.go:45]: This was added a long time ago. I think at that time using `KUBERNETES_VERSION` didn't work, but maybe things changed and let me try again.

Comment (pydctw) [reply test/e2e/suites/managed/eks_test.go:45]: Sounds great. This is not a blocker but a suggestion. If it takes too much time, we can refactor as a follow up.

Comment (richardchen331) [reply test/e2e/suites/managed/eks_test.go]: I removed it and now uses `KUBERNETES_VERSION` instead, it's tested and works well.

File: exp/api/v1beta1/awsmanagedmachinepool_webhook.go [hunk 21]
```diff
@@ -117,6 +117,26 @@ func (r *AWSManagedMachinePool) validateRemoteAccess() field.ErrorList {
...
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "InstanceType"), r.Spec.InstanceType, "InstanceType cannot be specified when LaunchTemplate is specified"))
+	}
+	if r.Spec.DiskSize != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "DiskSize"), r.Spec.DiskSize, "DiskSize cannot be specified when LaunchTemplate is specified"))
+	}
+
+	if r.Spec.AWSLaunchTemplate.IamInstanceProfile != "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "AWSLaunchTemplate", "IamInstanceProfile"), r.Spec.AWSLaunchTemplate.IamInstanceProfile, "IAM instance profile in launch template is prohibited in EKS managed node group"))
+	}
+
```
[the remark this answers is not in the document]
Comment (richardchen331) [reply exp/api/v1beta1/awsmanagedmachinepool_webhook.go:136]: `AMIType` is still required when making calls to EKS using the [AWS SDK](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#CreateNodegroupInput). 
We could not allow people to specify `AMIType` and use hardcoded value of `CUSTOM` when `AWSLaunchTemplate` is specified, but I'm wondering if there would be use cases where user want to create a node group without using launch template.

**Comment by richardchen331** [comment]
@pydctw Thanks for all the feedbacks! I've updated the PR based on your comments. Could you take another look? Thanks!

**Comment by pydctw** [comment]
Thanks for addressing all the comments, @richardchen331. It looks good to me. 

Two remaining items that are related regarding the usage of launchtemplate for AWSManagedMachinePool. Let's wait for @richardcase's opinion on this.
* [comment1](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#issuecomment-1211373213)
* [comment2](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#discussion_r946006474)

FYI, we are trying to include the PR in the next release.

File: exp/api/v1beta1/awsmanagedmachinepool_webhook.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (richardcase) [reply exp/api/v1beta1/awsmanagedmachinepool_webhook.go:136]: @pydctw - i added this to the agenda for Monday's office hours to discuss.

**Comment by richardchen331** [comment]
> Thanks for addressing all the comments, @richardchen331. It looks good to me.
> 
> Two remaining items that are related regarding the usage of launchtemplate for AWSManagedMachinePool. Let's wait for @richardcase's opinion on this.
> 
> * [comment1](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#issuecomment-1211373213)
> * [comment2](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#discussion_r946006474)
> 
> FYI, we are trying to include the PR in the next release.

@richardcase @pydctw just want to follow up and see if we have made any decision about these two comments? Thanks!

**Comment by pydctw** [comment]
> @richardcase @pydctw just want to follow up and see if we have made any decision about these two comments? Thanks!

Thanks for the reminder., @richardchen331. I planned to update the PR after the office hrs but forgot.

As there was no strong opinion whether to 1) Deprecate the existing fields and only support LaunchTemplate or 2) Support both existing and new way (LaunchTemplate), I propose we proceed as following.
* Merge the PR as is - it looks good and all test passed. 
* Discuss two options in the [proposal](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3365/files#diff-d6d8565c707995a7204010e1acade42238594c2bcf88d13a6ed7bd316369942eR117).
* Depending on the decision, create a new PR to mark those fields as deprecated if we decided on option 1, otherwise do nothing.

**Comment by pydctw** [comment]
/lgtm

cc @richardcase for an approval

**Comment by Ankitasw** [comment]
Can be moved to next API change.
cc @richardcase
